crates/wasi/src/preview2/tcp.rs

```diff
@@ -46,113 +45,206 @@ pub(crate) enum HostTcpState {
-    ///
-    /// The socket must be in non-blocking mode.
-    pub fn from_tcp_stream(tcp_socket: cap_std::net::TcpStream) -> io::Result<Self> {
-        let fd = rustix::fd::OwnedFd::from(tcp_socket);
-        let tcp_socket = TcpListener::from(fd);
-
-        let tcp_socket = unsafe {
-            tokio::net::TcpStream::try_from(std::net::TcpStream::from_raw_socketlike(
-                tcp_socket.into_raw_socketlike(),
-            ))
-            .unwrap()
+            Ok(n) => n,
+
+            // Failing with `EWOULDBLOCK` is how we differentiate between a closed channel and no
+            // data to read right now.
+            Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => 0,
+
+            Err(e) => {
+                tracing::debug!("unexpected error on TcpReadStream read: {e:?}");
+                self.closed = true;
+                0
+            }
         };
 
-        Ok(Self {
-            inner: Arc::new(HostTcpSocketInner { tcp_socket }),
-            tcp_state: HostTcpState::Default,
-        })
+        buf.truncate(n);
+        Ok((buf.freeze(), self.stream_state()))
     }
 
-    pub fn tcp_socket(&self) -> &tokio::net::TcpStream {
-        self.inner.tcp_socket()
+    async fn ready(&mut self) -> Result<(), anyhow::Error> {
+        if self.closed {
+            return Ok(());
+        }
+        self.stream.readable().await?;
+        Ok(())
     }
+}
 
-    pub fn clone_inner(&self) -> Arc<HostTcpSocketInner> {
-        Arc::clone(&self.inner)
-    }
+const SOCKET_READY_SIZE: usize = 1024 * 1024 * 1024;
+
+pub(crate) struct TcpWriteStream {
+    stream: Arc<tokio::net::TcpStream>,
+    write_handle: Option<AbortOnDropJoinHandle<anyhow::Result<()>>>,
 }
 
-impl HostTcpSocketInner {
-    pub fn tcp_socket(&self) -> &tokio::net::TcpStream {
-        let tcp_socket = &self.tcp_socket;
+impl TcpWriteStream {
+    pub(crate) fn new(stream: Arc<tokio::net::TcpStream>) -> Self {
+        Self {
+            stream,
+            write_handle: None,
+        }
+    }
+
+    /// Write `bytes` in a background task, remembering the task handle for use in a future call to
+    /// `write_ready`
+    fn background_write(&mut self, mut bytes: bytes::Bytes) {
+        assert!(self.write_handle.is_none());
+
+        let stream = self.stream.clone();
+        self.write_handle.replace(
+            tokio::spawn(async move {
+                // Note: we are not using the AsyncWrite impl here, and instead using the TcpStream
+                // primitive try_write, which goes directly to attempt a write with mio. This has
+                // two advantages: 1. this operation takes a &TcpStream instead of a &mut TcpStream
+                // required to AsyncWrite, and 2. it eliminates any buffering in tokio we may need
+                // to flush.
+                while !bytes.is_empty() {
+                    stream.writable().await?;
+                    let n = stream.try_write(&bytes)?;
```

Comment:
Could the error be handled here to detect EWOULDLBOCK? I'm worried about spurious writable notifications from above causing the write here to immediately return "WOULDBLOCK", but being able to handle that

Reply:
thanks, done